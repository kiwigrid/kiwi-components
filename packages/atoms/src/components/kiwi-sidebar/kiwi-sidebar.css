kiwi-sidebar.offcanvas-right .sidebar-right {
  box-shadow: -1px 0px 10px #aaa;
}
