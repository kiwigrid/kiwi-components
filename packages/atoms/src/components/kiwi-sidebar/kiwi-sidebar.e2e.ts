import { newE2EPage } from '@stencil/core/testing';

describe('kiwi-sidebar', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent('<kiwi-sidebar></kiwi-sidebar>');

    const element = await page.find('kiwi-sidebar');
    expect(element).toHaveClass('hydrated');
  });
});
