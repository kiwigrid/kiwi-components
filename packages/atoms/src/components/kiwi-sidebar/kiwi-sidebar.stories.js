import { html } from 'lit-html';
import notes from './readme.md';

export default {
  title: 'KiwiComponents/Sidebar',
  parameters: { docs: { description: { component: notes } } },
  decorators: [],
};

export const basic = (args) => {
  const { visible, position } = args;

  return html`
    <div class="panel panel-default">
      <div class="panel-body">This is outside of the sidebar</div>
    </div>
    <kiwi-sidebar visible=${visible} position=${position}>
      <div class="panel panel-default">
        <div class="panel-body">This is inside of the sidebar</div>
      </div>
    </kiwi-sidebar>
  `;
};
basic.args = {
  visible: true,
};
basic.argTypes = {
  position: {
    control: {
      type: 'select',
      options: ['left', 'right'],
    },
    defaultValue: 'right',
  },
};
