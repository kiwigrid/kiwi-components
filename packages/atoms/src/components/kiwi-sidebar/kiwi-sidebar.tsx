import {
  Component,
  ComponentInterface,
  h,
  Host,
  JSX,
  Prop,
  EventEmitter,
  Event,
  Listen,
} from '@stencil/core';

@Component({
  tag: 'kiwi-sidebar',
  styleUrl: 'kiwi-sidebar.css',
  shadow: false,
})
export class KiwiSidebar implements ComponentInterface {
  /** The position of the sidebar */
  @Prop()
  position!: 'left' | 'right';

  /** Visibility of the sidebar */
  @Prop()
  visible!: boolean;

  /** Event signaling a backdrop click. */
  @Event()
  requestClose!: EventEmitter<void>;

  @Listen('keydown', { target: 'document' })
  handleCloseOnEscape(ev: KeyboardEvent): void {
    if (!this.visible) {
      return;
    }

    if (ev.key && ev.key.startsWith('Esc')) {
      this.requestClose.emit();
    }
  }

  render(): JSX.Element {
    const offcanvas = this.visible ? `offcanvas-${this.position}` : '';
    const sidebar = `sidebar-offcanvas sidebar-${this.position} bg-gray p-2`;
    return (
      <Host class={offcanvas}>
        <div
          aria-label="Close Sidebar"
          aria-hidden="true"
          class="nav-backdrop"
          style={{ zIndex: this.visible ? '30' : '-1' }}
          onClick={this.handleBackdropClick}
        ></div>
        <div class={sidebar} aria-hidden={this.visible ? 'false' : 'true'}>
          <div class="sidebar-inner">
            <slot />
          </div>
        </div>
      </Host>
    );
  }

  private handleBackdropClick = (): void => {
    this.requestClose.emit();
  };
}
