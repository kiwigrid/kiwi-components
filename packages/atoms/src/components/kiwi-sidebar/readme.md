# kiwi-sidebar



<!-- Auto Generated Below -->


## Properties

| Property                | Attribute  | Description                 | Type                | Default     |
| ----------------------- | ---------- | --------------------------- | ------------------- | ----------- |
| `position` _(required)_ | `position` | The position of the sidebar | `"left" \| "right"` | `undefined` |
| `visible` _(required)_  | `visible`  | Visibility of the sidebar   | `boolean`           | `undefined` |


## Events

| Event          | Description                       | Type                |
| -------------- | --------------------------------- | ------------------- |
| `requestClose` | Event signaling a backdrop click. | `CustomEvent<void>` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
