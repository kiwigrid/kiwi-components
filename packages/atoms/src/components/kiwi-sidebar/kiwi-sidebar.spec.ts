import { KiwiSidebar } from './kiwi-sidebar';
import { newSpecPage } from '@stencil/core/testing';

describe('kiwi-sidebar', () => {
  it('builds', () => {
    expect(new KiwiSidebar()).toBeTruthy();
  });

  it('renders', async () => {
    const sidebarLeft = await newSpecPage({
      components: [KiwiSidebar],
      html: `
        <kiwi-sidebar position="left" visible>
          <span>Content</span>
        </kiwi-sidebar>
      `,
    });

    expect(sidebarLeft.root).toMatchSnapshot('left');

    const sidebarRight = await newSpecPage({
      components: [KiwiSidebar],
      html: `
        <kiwi-sidebar position="right" visible>
          <span>Content</span>
        </kiwi-sidebar>
      `,
    });

    expect(sidebarRight.root).toMatchSnapshot('right');

    const sidebarClosed = await newSpecPage({
      components: [KiwiSidebar],
      html: `
        <kiwi-sidebar position="left">
          <span>Content</span>
        </kiwi-sidebar>
      `,
    });

    expect(sidebarClosed.root).toMatchSnapshot('closed');
  });

  it('handles backdrop click', async () => {
    const sidebar = await newSpecPage({
      components: [KiwiSidebar],
      html: `
        <kiwi-sidebar position="left" visible>
          <span>Content</span>
        </kiwi-sidebar>
      `,
    });

    const sidebarInst: KiwiSidebar = sidebar.rootInstance;
    const sidebarElem = sidebar.root;

    jest.spyOn(sidebarInst.requestClose, 'emit');

    const backdrop: HTMLDivElement = sidebarElem!.querySelector(
      '.nav-backdrop',
    ) as HTMLDivElement;
    backdrop.click();

    expect(sidebarInst.requestClose.emit).toBeCalled();
  });

  it('handles esc key', async () => {
    const sidebar = await newSpecPage({
      components: [KiwiSidebar],
      html: `
        <kiwi-sidebar position="left" visible>
          <span>Content</span>
        </kiwi-sidebar>
      `,
    });

    const sidebarInst: KiwiSidebar = sidebar.rootInstance;

    jest.spyOn(sidebarInst.requestClose, 'emit');

    document.dispatchEvent(
      new KeyboardEvent('keydown', {
        key: 'Esc',
      }),
    );

    expect(sidebarInst.requestClose.emit).toBeCalled();

    sidebarInst.visible = false;

    document.dispatchEvent(
      new KeyboardEvent('keydown', {
        key: 'Esc',
      }),
    );

    expect(sidebarInst.requestClose.emit).toBeCalledTimes(1);

    document.dispatchEvent(
      new KeyboardEvent('keydown', {
        key: 'A',
      }),
    );

    expect(sidebarInst.requestClose.emit).toBeCalledTimes(1);

    document.dispatchEvent(
      new KeyboardEvent('keydown', {
        key: undefined,
      }),
    );

    expect(sidebarInst.requestClose.emit).toBeCalledTimes(1);
  });
});
