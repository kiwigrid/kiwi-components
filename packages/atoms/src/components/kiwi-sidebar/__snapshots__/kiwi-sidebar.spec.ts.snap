// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kiwi-sidebar renders: closed 1`] = `
<kiwi-sidebar position="left">
  <!---->
  <div aria-hidden="true" aria-label="Close Sidebar" class="nav-backdrop" style="z-index: -1;"></div>
  <div aria-hidden="true" class="bg-gray p-2 sidebar-left sidebar-offcanvas">
    <div class="sidebar-inner">
      <span>
        Content
      </span>
    </div>
  </div>
</kiwi-sidebar>
`;

exports[`kiwi-sidebar renders: left 1`] = `
<kiwi-sidebar class="offcanvas-left" position="left" visible="">
  <!---->
  <div aria-hidden="true" aria-label="Close Sidebar" class="nav-backdrop" style="z-index: 30;"></div>
  <div aria-hidden="false" class="bg-gray p-2 sidebar-left sidebar-offcanvas">
    <div class="sidebar-inner">
      <span>
        Content
      </span>
    </div>
  </div>
</kiwi-sidebar>
`;

exports[`kiwi-sidebar renders: right 1`] = `
<kiwi-sidebar class="offcanvas-right" position="right" visible="">
  <!---->
  <div aria-hidden="true" aria-label="Close Sidebar" class="nav-backdrop" style="z-index: 30;"></div>
  <div aria-hidden="false" class="bg-gray p-2 sidebar-offcanvas sidebar-right">
    <div class="sidebar-inner">
      <span>
        Content
      </span>
    </div>
  </div>
</kiwi-sidebar>
`;
